### PR TITLE
datetime: redo for full features

### PR DESCRIPTION
**Pull request by gigagrug**
## Description
1. Add Unixepoch modifier
2. Add JulianDay  modifier
3. Add Auto  modifier
4. Fix Localtime and UTC shouldn't double convert
5. fix err bug: returns "" needs to return null `SELECT typeof(strftime('%Y', 'not-a-valid-date'));`
6. add support for decimal month and year calculations (on top of jussi's added support for days, hours, min, sec)
7. fix the bugs @LeMikaelF found in my code from https://github.com/tursodatabase/turso/pull/4104.
https://github.com/tursodatabase/turso/issues/602

## Motivation and context
I originally had this pr: https://github.com/tursodatabase/turso/pull/4104
But then I saw: https://github.com/tursodatabase/turso/pull/4345 that did things better.
However although my code uses some code in the PR I couldn't add all the code without breaking compatibility. Specifically the faster parser.

There is also this PR: https://github.com/tursodatabase/turso/pull/4331
Storing date in struct is closer to how SQLite handles date but I personally found too much reliance on the Chronos create a lot of bugs with compatibility.

## Description of AI Usage

I used Google Gemini Pro.
1. have it generate tests first.
2. then verify tests in sqlite and fix the mistakes.
3. generate code based on tests.
4. run tests.
5. tell Gemini the failed tests.

### CODE REVIEW

**Comment by gigagrug** [comment]
@LeMikaelF I know you're busy but can you help me with this. I have no idea how my code would cause this error:
`FAIL [   0.628s] ( 999/1605) turso_core mvcc::database::tests::test_cursor_with_btree_and_mvcc_fuzz`

**Comment by LeMikaelF** [comment]
> @LeMikaelF I know you're busy but can you help me with this. I have no idea how my code would cause this error: `FAIL [ 0.628s] ( 999/1605) turso_core mvcc::database::tests::test_cursor_with_btree_and_mvcc_fuzz`

You were just (unlucky) enough that your CI run found a failing seed in one of the fuzz tests. I've opened an issue https://github.com/tursodatabase/turso/issues/4371 and restarted the CI action. It should be good now.

Don't hesitate to tag us in for things like this.

**Comment by gigagrug** [comment]
@LeMikaelF seem to be two more fuzz tests failures. 
Also I was wondering if you can take a look at this (WIP): https://github.com/gigagrug/turso/blob/raw_datetime/core/functions/datetime.rs 
The reason that decided to port date.c was because I kept running into to edge cases around the fact that SQLite allows some invalid datetimes but chrono doesn't seem to allow any. I think the heavy use of chrono will result in never ending edge cases. This also reduces reliance of a third-party dependency which is closer to how SQLite does things. 
However I don't know if this will make it incompatible with Turso's features like concurrent writes. Let me know if you think this is good direction or not.
Screenshot 2026-01-01 160843

**Comment by LeMikaelF** [comment]
Sorry for the delay, I was away on vacation. I'll take a look tomorrow.

File: core/functions/datetime.rs [hunk 1]
```diff
@@ -63,13 +63,32 @@ where
     exec_datetime(values, DateTimeOutput::StrfTime(format_str))
 }
 
+pub fn exec_julianday<I, E, V>(values: I) -> Value
+where
+    V: AsValueRef,
+    E: ExactSizeIterator<Item = V>,
+    I: IntoIterator<IntoIter = E, Item = V>,
+{
+    exec_datetime(values, DateTimeOutput::JuliaDay)
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: typo: `JuliaDay` -> `JulianDay`, I think it was like this before, but could you fix it please?

File: core/functions/datetime.rs [hunk 2]
```diff
@@ -80,57 +99,132 @@ where
 {
     let values = values.into_iter();
     if values.len() == 0 {
-        let now = chrono::Local::now().to_utc().naive_utc();
+        let (now, _is_utc, _ov) = parse_naive_date_time(Value::build_text("now")).unwrap();
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: I'm curious why this changed, it will trim and compare the slice `now` and always return `(chrono::Local::now().to_utc().naive_utc(), true, 0)`, any reason why you added the indirection?

At line 139, it's still the bare `chrono::Local::now().to_utc().naive_utc()` that's used.

File: core/functions/datetime.rs [hunk 3]
```diff
@@ -80,57 +99,132 @@ where
...
+        let parsed = parse_naive_date_time(first);
+
+        let is_mod = if let ValueRef::Text(s) = first.as_value_ref() {
+            let s = s.as_str();
+            s.eq_ignore_ascii_case("subsec") || s.eq_ignore_ascii_case("subsecond")
+        } else {
+            false
+        };
+
+        (initial_numeric, is_mod, parsed)
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: At this point of reading the PR, I don't understand why `is_mod` is aliased to `is_implicit_now`. Maybe find a better name?

File: core/functions/datetime.rs [hunk 4]
```diff
@@ -80,57 +99,132 @@ where
...
+
+    if let Some((mut dt, is_utc, overflow)) = parsed_result {
+        modify_dt(
+            &mut dt,
+            values.skip(1),
+            output_type,
+            initial_numeric_value,
+            true,
+            is_utc,
+            overflow, // Pass initial overflow!
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: unless there's a reason that I missed, this comment seems superfluous

```suggestion
            overflow,
```

File: core/functions/datetime.rs [hunk 5]
```diff
@@ -80,57 +99,132 @@ where
...
                 Ok(true) => subsec_requested = true,
                 Ok(false) => {}
-                Err(_) => return Value::build_text(""),
+                Err(_) => return Value::Null,
             }
 
             if matches!(parsed, Ok(Modifier::Floor) | Ok(Modifier::Ceiling)) {
                 n_floor = 0;
             }
+            is_first_modifier = false;
```
Comment (LeMikaelF) [inline core/functions/datetime.rs:225]: this is not used

```suggestion
```

File: core/functions/datetime.rs [hunk 6]
```diff
@@ -246,28 +390,99 @@ fn apply_modifier(dt: &mut NaiveDateTime, modifier: &str, n_floor: &mut i64) ->
...
+                let total_ms = (val * 1000.0).round() as i64;
+
+                if let Some(base) = DateTime::from_timestamp(0, 0) {
+                    // TimeDelta handles the negative math (e.g. -500ms correctly goes back 1 second)
+                    match base.checked_add_signed(TimeDelta::milliseconds(total_ms)) {
+                        Some(new_dt) => *dt = new_dt.naive_utc(),
+                        None => return Err(InvalidModifier("Date overflow".into())),
+                    }
+                } else {
+                    return Err(InvalidModifier("Internal error".into()));
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: since line 408 will never return `None`, could you `expect` it? same for line 451.

File: testing/scalar-functions-datetime.test [hunk 7]
```diff
@@ -235,6 +284,130 @@ do_execsql_test unixepoch-at-millisecond-precision-input-produces-seconds-precis
   SELECT unixepoch('2022-01-27 12:59:28.052');
 } {1643288368}
 
+do_execsql_test unixepoch-with-modifier-start-of-day {
```
Comment (LeMikaelF) [inline testing/scalar-functions-datetime.test:401]: There's a panic that happens on this branch but not on main:

```
turso> SELECT datetime(-1, 'julianday');

thread 'main' panicked at core/functions/datetime.rs:946:17:
attempt to multiply with overflow
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: testing/scalar-functions-datetime.test [hunk 8]
```diff
@@ -95,6 +95,30 @@ do_execsql_test date-with-invalid-modifier {
   SELECT date('2023-05-18', 'invalid modifier');
 } {{}}
 
+do_execsql_test date-with-auto-julianday-top {
```
Comment (LeMikaelF) [inline testing/scalar-functions-datetime.test:212]: here are some missing tests, identified by Claude. I used this prompt:

> look at this PR, and verify that all of the new behaviour is tested in scalar-functions-datetime.test . To do this, split the implementation code into functions, and give each one to a sub-agent that will walk through all branches, explicit and implicit, and match them with the new tests in scalar-functions-datetime.test.

```tcl

# julianday edge cases

do_execsql_test julianday-null-input {
  SELECT julianday(NULL);
} {{}}

do_execsql_test julianday-invalid-input {
  SELECT julianday('not a date');
} {{}}

do_execsql_test julianday-empty-string {
  SELECT julianday('');
} {{}}

do_execsql_test julianday-negative-timezone {
  SELECT julianday('2023-05-18 15:30:45-05:00');
} {2460083.3546875}

do_execsql_test julianday-z-suffix {
  SELECT julianday('2023-05-18 15:30:45Z');
} {2460083.14635417}

do_execsql_test julianday-iso8601-t-separator {
  SELECT julianday('2023-05-18T15:30:45');
} {2460083.14635417}

do_execsql_test julianday-out-of-range {
  SELECT julianday('10001-01-01');
} {{}}

do_execsql_test julianday-string-with-modifier {
  SELECT julianday('2023-05-18 12:00:00', '+1 day');
} {2460084.0}

do_execsql_test julianday-string-with-start-of-month {
  SELECT julianday('2023-05-18', 'start of month');
} {2460065.5}

do_execsql_test julianday-string-with-minutes-modifier {
  SELECT julianday('2023-05-18 12:00:00', '+30 minutes');
} {2460083.02083333}

# timediff argument count

do_execsql_test_error timediff-one-arg {
  SELECT timediff('12:00:00');
} {wrong number of arguments.*}

# floor/ceiling edge cases

do_execsql_test date-floor-then-add-day {
  SELECT date('2024-01-31', '+1 month', 'floor', '+1 day');
} {2024-03-01}

do_execsql_test datetime-floor-then-add-hours {
  SELECT datetime('2024-01-31 10:00:00', '+1 month', 'floor', '+2 hours');
} {{2024-02-29 12:00:00}}

do_execsql_test date-fractional-month-floor {
  SELECT date('2024-01-15', '+1.5 month', 'floor');
} {2024-03-01}

do_execsql_test datetime-fractional-month-floor {
  SELECT datetime('2024-01-15 00:00:00', '+1.5 month', 'floor');
} {{2024-03-01 00:00:00}}

# modifier ordering errors

do_execsql_test unixepoch-not-first-modifier {
  SELECT date(1000, '+1 day', 'unixepoch');
} {{}}

do_execsql_test julianday-not-first-modifier {
  SELECT date(2460000, '+1 day', 'julianday');
} {{}}

do_execsql_test unixepoch-after-start-of-day {
  SELECT datetime(86400, 'start of day', 'unixepoch');
} {{}}

do_execsql_test julianday-after-hours-modifier {
  SELECT time(2460083, '+1 hour', 'julianday');
} {{}}

# negative date-only offset

do_execsql_test date-negative-date-offset {
  SELECT date('2023-05-18', '-0001-01-01');
} {2022-04-17}

do_execsql_test date-negative-date-offset-2 {
  SELECT date('2023-12-31', '-0001-00-00');
} {2022-12-31}

do_execsql_test datetime-negative-date-offset {
  SELECT datetime('2023-05-18 12:30:00', '-0001-01-01');
} {{2022-04-17 12:30:00}}

do_execsql_test unixepoch-negative-date-offset {
  SELECT unixepoch('2023-05-18', '-0001-01-01');
} {1650153600}
```

File: core/functions/datetime.rs [hunk 9]
```diff
@@ -478,64 +580,323 @@ fn get_date_time_from_time_value_string(value: &str) -> Option<NaiveDateTime> {
...
+fn get_date_time_from_time_value_float(value: f64) -> Option<(NaiveDateTime, bool, i64)> {
     if value.is_infinite() || value.is_nan() || !is_julian_day_value(value) {
         return None;
     }
-    julian_day_to_datetime(value).ok()
+    // 0 Overflow
+    julian_day_to_datetime(value).ok().map(|dt| (dt, true, 0))
+}
+
+fn parse_naive_date_time(time_value: impl AsValueRef) -> Option<(NaiveDateTime, bool, i64)> {
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: The `(NaiveDateTime, bool, i64)` type is not very expressive. Without having read its usages, it's hard to know what the `bool` and the `i64` represent. For example, just by reading `get_date_time_from_time_value_float()`, it's hard to know what `true` and `0` are. It would probably be more expressive as a struct, so that the fields can be named. What do you think?

File: core/functions/datetime.rs [hunk 10]
```diff
@@ -478,64 +580,323 @@ fn get_date_time_from_time_value_string(value: &str) -> Option<NaiveDateTime> {
...
+                && m_part.chars().all(|x| x.is_ascii_digit())
+                && is_valid_tz(h_part, m_part)
+            {
+                return &s[..idx];
+            }
+        }
+    }
+
+    // Check +HHMM or -HHMM (length >= 5)
+    if s.len() >= 5 {
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: SQLite doesn't seem to accept the 5-char and 3-char formats:

```
sqlite> SELECT date('2023-05-18 23:30:45+1200');

sqlite> SELECT date('2023-05-18 23:30:45+12');


```

I suggest adding these as tests to the test suite.

File: core/functions/datetime.rs [hunk 11]
```diff
@@ -478,64 +580,323 @@ fn get_date_time_from_time_value_string(value: &str) -> Option<NaiveDateTime> {
...
+        dt.month() as i32,
+        dt.day() as i32,
+        dt.hour() as i32,
+        dt.minute() as i32,
+        s,
+    )
+}
+// Helper to compute JD from raw components (allows invalid dates like 2023-02-31)
+// Helper to compute JD from raw components
+// 's' is now a double that includes subseconds (e.g., 12.345)
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]:
```suggestion
// 's' is a double that includes subseconds (e.g., 12.345)
```

File: core/functions/datetime.rs [hunk 12]
```diff
@@ -1000,7 +1394,7 @@ mod tests {
     fn test_invalid_get_date_from_time_value() {
         let invalid_cases = vec![
             Value::build_text("2024-07-21 25:00"),    // Invalid hour
-            Value::build_text("2024-07-21 24:00:00"), // Invalid hour
+            Value::build_text("2024-07-21 25:00:00"), // Invalid hour
```
Comment (LeMikaelF) [inline core/functions/datetime.rs:1400]: `select date('2024-07-21 24:00:00');` used to incorrectly return `NULL`, but now it returns the wrong date. I created https://github.com/tursodatabase/turso/issues/4424 so that we can track this issue.

File: core/functions/datetime.rs [hunk 13]
```diff
@@ -1903,9 +2446,9 @@ mod tests {
         let expected = Value::Null;
         assert_eq!(exec_timediff(&[start, end.clone()]), expected);
 
-        // Test identical times - should return zero duration, not Null
```
Comment (LeMikaelF) [inline core/functions/datetime.rs:1906]: Any reason you removed this test? It should probably stay.

File: core/functions/datetime.rs [hunk 14]
```diff
@@ -2026,4 +2569,105 @@ mod tests {
...
+    fn test_unixepoch_numeric_modifiers_julianday() {
+        let result = exec_unixepoch(&[Value::Float(2440587.5), text("julianday")]);
+        assert_eq!(result, Value::Integer(0));
+
+        let result = exec_unixepoch(&[Value::Float(2460311.5), text("julianday")]);
+        assert_eq!(result, Value::Integer(1704153600));
+
+        let result = exec_unixepoch(&[Value::Float(0.0), text("julianday")]);
+        match result {
+            Value::Integer(i) => assert!(i < -210866700000, "Expected very old timestamp, got {i}"),
```
Comment (LeMikaelF) [inline core/functions/datetime.rs]: unless there was a reason for the `<` assertion, this is more precise:

```suggestion
            Value::Integer(i) => assert_eq!(i, -210866760000),
```

**Comment by gigagrug** [comment]
@LeMikaelF Thank you so much for your time! I did not know about this `select distinct * from (select datetime('subsecond') from generate_series(1, 100));` and for finding my code issues.

But you seemed to miss that this pr is different from the link that I shared. That link is a different branch to removes the use of Chrono NaiveDatetime. This is because chrono does not seem to allow raw numbers like sqlite does. This is why `24:59:59` fails also others. This branch fixes this: https://github.com/gigagrug/turso/blob/raw_datetime/core/functions/datetime.rs but is complete different approach ports the code from: https://github.com/sqlite/sqlite/blob/master/src/date.c . 

If you want me to abandon raw_datetime then just let me know and I'll fix my code on this pr.

**Comment by LeMikaelF** [comment]
@gigagrug I'm not sure I follow, the SQL you quote is from an issue I opened while I was reviewing this PR because it made me notice the issue (https://github.com/tursodatabase/turso/issues/4428), but it was already present before this PR. It's not directly related to this PR.

I haven't reviewed the other branch, do you mean that you would rather merge the other branch and not this one?

Sorry if I'm missing something obvious, it's getting late here.

File: core/functions/datetime.rs (hunk 6 above)
Comment (gigagrug) [reply core/functions/datetime.rs]: using `expect` seem to panic on large numbers when it should null
ex. `select datetime(9466848000000.0, 'unixepoch');`
I switch to using `ok_or_else` flattens the code.

File: core/functions/datetime.rs (hunk 5 above)
Comment (gigagrug) [reply core/functions/datetime.rs:225]: take for example: `select unixepoch(0, 'unixepoch', 'julianday');` should return null.
if i remove `is_first_modifier = false;` it will return: -210866760000. applying julianday.
if `select unixepoch(0, 'julianday', 'unixepoch');` returns 0 applying unixepoch.

As Claude has properly identified when you asked it to generate missing tests modifier ordering matters. Removing this makes them all valid when they should be NULL.

Comment (LeMikaelF) [reply core/functions/datetime.rs:225]: I misread the code, sorry!

**Review by LeMikaelF — APPROVED** [review]
LGTM, about the raw_datetime branch, feel free to open it as another PR. There's an issue to track the minor incompatibility with hour 24 (https://github.com/tursodatabase/turso/issues/4424), but I don't see it as something significant.

Thanks for all the work!
